components/script/dom/htmlimageelement.rs

```diff
@@ -871,6 +871,166 @@ impl HTMLImageElement {
+        let mut selected_pixel_density = None;
+        let elem = self.upcast::<Element>();
+        let document = document_from_node(elem);
+        let base_url = document.base_url();
+        match self.select_image_source() {
+            Some(src) => {
+                // Step 3
+                selected_source = Some(src.0);
+                selected_pixel_density = Some(src.1);
+            },
+            _ => ()
+        }
+
+        // Step 4
+        let src = match selected_source {
+            Some(src) => src,
+            _ => return,
+        };
+
+        // Step 5
+        let same_source = match *self.last_selected_source.borrow() {
+            Some(ref last_src) => *last_src == src,
+            _ => false,
+        };
+
+        let same_selected_pixel_density = match selected_pixel_density {
+            Some(den) => self.current_request.borrow().current_pixel_density.map_or(false, |density| density == den),
+            _ => self.current_request.borrow().current_pixel_density.map_or(true, |_den| false),
+        };
+
+        if same_source && same_selected_pixel_density {
+            return;
+        }
+
+        // Step 6
+        if let Ok(img_url) = base_url.join(&String::from(src.clone())) {
+            // Step 12
+            *self.pending_request.borrow_mut() = ImageRequest {
+                state: State::Unavailable,
+                parsed_url: Some(img_url.clone()),
+                source_url: None,
+                image: None,
+                metadata: None,
+                blocker: None,
+                final_url: None,
+                current_pixel_density: None,
+            };
+
+            let window = window_from_node(self);
+            let image_cache = window.image_cache();
+            // Step 14
+            let response =
+                image_cache.find_image_or_metadata(img_url.clone().into(),
+                                                   UsePlaceholder::No,
+                                                   CanRequestImages::No);
+            if let Ok(ImageOrMetadataAvailable::ImageAvailable(_image, _url)) = response {} else {
+                let window = window_from_node(self);
+                let image_cache = window.image_cache();
+                // Step 14.1
+                let response =
+                    image_cache.find_image_or_metadata(img_url.clone().into(),
+                                                       UsePlaceholder::Yes,
+                                                       CanRequestImages::Yes);
+                match response {
+                    // TODO Handle multipart/x-mixtape
+                    Ok(ImageOrMetadataAvailable::ImageAvailable(_image, _url)) => {}
+                    Ok(ImageOrMetadataAvailable::MetadataAvailable(_m)) => { }
+                    Err(ImageState::Pending(_id)) => { }
+                    Err(ImageState::NotRequested(_id)) => { }
+                    Err(ImageState::LoadError) => {
```

Comment:
So there's some work that needs to happen here in order to fetch the image and deal with the response before we proceed to step 15. I would start by duplicating `fetch_image` as `fetch_image_for_environment_change`, as well as `process_image_response` as `process_image_response_for_environment_change`. The new method for processing an image response would be less complex, because it only needs to deal with the case where an image is fully loaded or unavailable for any reason (step 14.5). Once the image response is complete, we can call the `finish_reacting_to_environment_change` method that I suggested earlier.